src/test/spec/unified_runner/mod.rs

```diff
@@ -0,0 +1,304 @@
+mod entity;
+mod matcher;
+mod operation;
+mod test_event;
+mod test_file;
+
+use std::collections::HashMap;
+
+use futures::stream::TryStreamExt;
+use lazy_static::lazy_static;
+use semver::Version;
+
+use crate::{
+    bson::{doc, Document},
+    concern::{Acknowledgment, WriteConcern},
+    options::{
+        CollectionOptions,
+        FindOptions,
+        InsertManyOptions,
+        ReadConcern,
+        ReadPreference,
+        SelectionCriteria,
+    },
+    test::{run_spec_test, util::EventClient, TestClient, LOCK},
+};
+
+pub use self::{
+    entity::Entity,
+    matcher::results_match,
+    operation::EntityOperation,
+    test_event::TestEvent,
+    test_file::{CollectionData, Entity as TestFileEntity, ExpectError, Operation, TestFile},
+};
+
+lazy_static! {
+    static ref SPEC_VERSIONS: Vec<Version> = vec![Version::parse("1.0.0").unwrap()];
+}
+
+pub async fn run_unified_format_test(test_file: TestFile) {
+    let version_matches = SPEC_VERSIONS.iter().any(|req| {
+        if req.major != test_file.schema_version.major {
+            return false;
+        }
+        if req.minor < test_file.schema_version.minor {
+            return false;
+        }
+        // patch versions do not affect the test file format
+        true
+    });
+    if !version_matches {
+        println!(
+            "Test runner not compatible with specification version {}",
+            &test_file.schema_version
+        );
+        return;
+    }
+
+    let client = TestClient::new().await;
+    if let Some(requirements) = test_file.run_on_requirements {
+        if !requirements
+            .iter()
+            .any(|requirement| requirement.can_run_on(&client))
+        {
+            println!("Client topology not compatible with test");
+            return;
+        }
+    }
+
+    for test_case in test_file.tests {
+        if let Some(skip_reason) = test_case.skip_reason {
+            println!("skipping {}: {}", &test_case.description, skip_reason);
+            return;
+        }
+
+        if let Some(requirements) = test_case.run_on_requirements {
+            if !requirements
+                .iter()
+                .any(|requirement| requirement.can_run_on(&client))
+            {
+                println!(
+                    "{}: client topology not compatible with test",
+                    &test_case.description
+                );
+                return;
+            }
+        }
+
+        if let Some(ref initial_data) = test_file.initial_data {
+            for data in initial_data {
+                insert_initial_data(data, &client).await;
+            }
+        }
+
+        let mut entities: HashMap<String, Entity> = HashMap::new();
+        if let Some(ref create_entities) = test_file.create_entities {
+            populate_entity_map(&mut entities, create_entities).await;
+        }
+
+        let mut fail_points: Vec<Document> = Vec::new();
```

Comment:
Could we name this something that indicates that this contains the commands to disable the failpoints, e.g. `failpoint_disable_commands`?